### ISSUE TITLE
Assembler Example

### ISSUE THREAD

**Issue by pbelmann** [bioboxes/rfc#90]
We should test the specs defined in #61 by building two working examples:

### Ray Assembler
Signature:
`ray:[(Id,fastq)] -> fasta` 

More complex example:
### Soap Assembler
Signature:
`soap:[(Id,fastq)],[(Id,fragment_size)] -> fasta`
Any definition will be checked against a schema discussed in #86.

They should accept a yaml like:
```YAML
---
arguments:
     fastq:
      - path: "/path/inside/the/container/to/lib1"
        id: "pe"
        type: paired
      - path: "/path/inside/the/container/to/lib2"
        id: "se"
        type: single
      - path: "/path/inside/the/container/to/lib2"
        id: "mp"
        type: paired
     fragment_size:
      - id : "pe",
        value: 240
      - id : "mp",
        value: 5000
```

The current rx based schema for assemblers is:

```YAML
---
  type: "//rec"
  required: 
    version: "//str"
    arguments: 
      type: "//rec"
      required: 
        fastq: 
          type: "//arr"
          contents: 
            type: "//rec"
            required: 
              path: "//str"
              id: "//str"
              type: 
                type: "//any"
                of: 
                  - 
                    type: "//str"
                    value: "paired"
                  - 
                    type: "//str"
                    value: "single"
      optional: 
        fragment_size: 
          type: "//arr"
          contents: 
            type: "//rec"
            required: 
              id: "//str"
              value: "//num"
```

I will create a repository for ray and soap in bioboxes. You can comment the progress in the corresponding respositories.
#### Please note that this is maybe not the final spec since we are still discussing it in the issues #61 and #86. On any progress in that issues I will update this issue and the repositories.

**Comment by pbelmann** [comment]
I created the ray repository that works with the above schema. @bioboxes/members have write access.

@michaelbarton  mentioned in #61 that an assembler should have a yaml as output that contains the path to the fasta. I think this addresses also #9 so I suggest the following example yaml:

```YAML
---
version: 0.9.0
arguments:
  - value: "relative/path/to/contigs/in/mounted/output/dir"
    type: fasta
```
and schema:

```YAML
---
  type: "//rec"
  required: 
    version: "//str"
    arguments: 
      type: "//arr"
      contents: 
        type: "//any"
        of: 
          - 
            type: "//rec"
            required: 
              type: 
                type: "//str"
                value: "fasta"
              value: "//str"

```

**Comment by michaelbarton** [comment]
Thanks for creating the example assembler @pbelmann. You mention the signature
at the start of the README. I think since the signature is also structured
data, the signatures could be defined as YAML too. For example if the signature
was:

  soap: [fastq A], [fragment_size A] -> fasta B

Then the YAML signature would be:

YAML```
---
version: X.Y.Z
input:
  - [{fastq: A}]
  - [{fragment_size: A}]
output:
  - {fasta: B}
```

The input key has two values corresponding to the two arguments. Each of these
arguments is also a list: the fastqs and the fragment_sizes, and so is defined
in the YAML as so. I thought that if we used this format, the input file
validator could be generated from this spec and we wouldn't need to write the
validator by hand.

Part of the reason of bioboxes is standardise the interfaces so perhaps we
should also consider limiting the number of signatures to begin with. These
could be:

    f: [fastq A]                                 -> fasta B
    f: [paired_fastq A], [fragment_size A]       -> fasta B
    f: [paired_fastq A], [Maybe fragment_size A] -> fasta B

A limited number of signatures would limit the amount of work to create a
working spec. I know that time is an issue for CAMI and you like to be able to
ask others to start submitting bioboxes. To start with, we could write the
validators for these three types by hand, and postpone automatically generating
the validators.

**Comment by pbelmann** [comment]
I created the ray repository that works with the schema of my last comment in this issue. @bioboxes/members have write access.

>I think since the signature is also structured
>data, the signatures could be defined as YAML too. 

I agree. I will create the issues shortly for soap and ray.

>The input key has two values corresponding to the two arguments. Each of these
>arguments is also a list: the fastqs and the fragment_sizes, and so is defined
>in the YAML as so. I thought that if we used this format, the input file
>validator could be generated from this spec and we wouldn't need to write the
>validator by hand.

I agree. It think it is even possible to create a custom type in schema library like rx(what we are using at the moment (see #86 )).  That way we could mix the types according to the signature. If it works we don't have to write any code for this. It would alway be something like (python example):
```PYTHON
rx = Rx.Factory({ "register_core_types": True })
schema = rx.make_schema(yaml.load(open(schema_file)))
schema.check(data)
```
>Part of the reason of bioboxes is standardise the interfaces so perhaps we
>should also consider limiting the number of signatures to begin with. These
>could be:

>f: [fastq A]                                 -> fasta B
>f: [paired_fastq A], [fragment_size A]       -> fasta B
>f: [paired_fastq A], [Maybe fragment_size A] -> fasta B

Should we really define every possible signature? I think it would be to many combinations in later bioboxes versions?
What do you think about just defining the types in the spec and an assembler returns you the signatures it accepts?

>A limited number of signatures would limit the amount of work to create a
>working spec. I know that time is an issue for CAMI and you like to be able to
>ask others to start submitting bioboxes. To start with, we could write the
>validators for these three types by hand, and postpone automatically 
>generating the validators.

It is not only for CAMI. When we can solve issues #61, #90, #86 and maybe #85 we are a really big step closer to #9.   

What do mean by automatically generating validators? 
I agree we should have vaildators and I would start to build a validator like you already did (https://github.com/bioboxes/validator-short-read-assember) and it would to to begin with just scheck if an assembler accepts
* a yaml according to a schema and
* produces the input and ouput data

**Comment by michaelbarton** [comment]
> I agree. It think it is even possible to create a custom type in schema
> library like rx(what we are using at the moment (see #86 )). That way we
> could mix the types according to the signature. If it works we don't have to
> write any code for this. It would alway be something like (python example):

I wasn't too clear on this. Do you mean we could generate an Rx document from a
YAML-serialised signature?

> Should we really define every possible signature? I think it would be to many
> combinations in later bioboxes versions? What do you think about just
> defining the types in the spec and an assembler returns you the signatures it
> accepts?

I don't think we should define every signature, instead I thought it might be
useful to require certain signatures are available in each biobox type. For
example if every short read assembler provided this signature as a minimum:

    f: [paired_fastq A], [Maybe fragment_size A] -> fasta B

Then it would always be possible to benchmark a new published assembler biobox
against existing ones because they always implement at least one signature in
common. I agree that we should have a limited number of types, like a
controlled vocabulary. We could take these from EDAM. I think to begin with the
short read assembler, we'll just need terms for fastq types and fragment sizes.

> What do mean by automatically generating validators?

There will be two types of validators:

  * input YAML file validators
  * biobox validators

I was referring to automatically generating a YAML file validator from the
signature which I think you mentioned already. The biobox validators I think
we'll still have to write ourselves. Though if everything is structured data we
could eventually automatically generate these ourselves too.

**Comment by pbelmann** [comment]
>I wasn't too clear on this. Do you mean we could generate an Rx document from a
>YAML-serialised signature?

In my opinion there are two scenarios where a schema could be useful (it does not have to be rx).

### Schema Scenarios

1. We could create one rx schema that checks all possible definitions for an assembler yaml.
This is what I'm doing at the moment. 
Example:
When you take a look at my first comment in this issue, it accepts the signatures `[fastq],[fragment_size] -> X` and  `[fastq] -> X` and nothing else. Even the definition for an fastq entry is defined in the schema.   
The advantage here is that we could create a base image with a schema validator, so a developer of biobox would just have to use our base image.

2. We could create a schema for every possible signature (maybe a repository in bioboxes). The developer could choose his schema and just call two lines of code for every possible schema in python, ruby, perl or JavaScript.
Example:
So we would have to define one yaml file schema for  `[fastq],[fragment_size] -> X` and one schema for 
`[fastq] -> X`.

### Important points

* In both cases we have to discuss in #86 whether rx is the right choice because while it is concise it just gives you feedback if a schema is valid or not and not what is actually wrong with the input yaml file (Please note I just looked at the python implementation).

* In both cases we have to create our own custom types in schema (In think this is possible in any schema libraray). At the moment I'm using the [core types](http://rx.codesimply.com/coretypes.html#single) in rx. But you can create your own [custom types](http://rx.codesimply.com/moretypes.html) in separate files. That way we don't have to write a complex schema like that in my first comment but define and name every subschema and just referencing it in other schemas.

But I think we should continue the schema discussion in issue #86.

>I don't think we should define every signature, instead I thought it might be
>useful to require certain signatures are available in each biobox type. For
>example if every short read assembler provided this signature as a minimum:

>f: [paired_fastq A], [Maybe fragment_size A] -> fasta B

>Then it would always be possible to benchmark a new published assembler biobox
>against existing ones because they always implement at least one signature in
>common. I agree that we should have a limited number of types, like a
>controlled vocabulary. We could take these from EDAM. I think to begin with the
>short read assembler, we'll just need terms for fastq types and fragment sizes.

I agree. I think we should reuse existing ontologies like EDAM as much as possible. I will take a look in EDAM before starting to define a first draft for the new assembler spec.

**Comment by pbelmann** [comment]
In issue #61 we decided on a YAML like the following:

```YAML
---
version: 0.9.0
arguments:
    - fastq:
      - path: "/path/inside/the/container/to/lib1"
        id: "pe"
        type: paired
      - path: "/path/inside/the/container/to/lib2"
        id: "se"
        type: single
      - path: "/path/inside/the/container/to/lib2"
        id: "mp"
        type: paired
    - fragment_size:
      - id : "pe",
        value: 240
      - id : "mp",
        value: 5000
```

For now I choose the rx schema until we decide on maybe another library in #86 : 

```YAML
---
type: "//rec"
required: 
    version: "//str"
    arguments: 
      type: "//arr"
      length: 
        min: 1
      contents: 
        type: "//any"
        of: 
          - 
            type: "//rec"
            required: 
              fastq: 
                type: "//arr"
                contents: 
                  type: "//rec"
                  required: 
                    path: "//str"
                    id: "//str"
                    type: 
                      type: "//any"
                      of: 
                        - 
                          type: "//str"
                          value: "paired"
                        - 
                          type: "//str"
                          value: "single"
                length: 
                  min: 1
          - 
            type: "//rec"
            required: 
              fragment_size: 
                type: "//arr"
                contents: 
                  type: "//rec"
                  required: 
                    id: "//str"
                    value: "//num"
                length: 
                  min: 1
```

**Comment by pbelmann** [comment]
The output YAML file should have the output: 

---
version: 0.9.0
arguments:
  - value: STRING
    type: CONTIG or SCAFFOLD  

Value is the relative path to the CONTIG or SCAFFOLD fasta file.

**Comment by michaelbarton** [comment]
@pbelmann If the input and output formats should match, should it look something like?

```YAML
---
version: 0.9.0
arguments:
  - fasta:
    id: auto_generated_id_1
    value: "/path/to/reads"
    type: contig
```

**Comment by pbelmann** [comment]
I agree. Could you create the issues, I will solve this tomorrow.

**Comment by michaelbarton** [comment]
Ok. Created issue #98

**Comment by pbelmann** [comment]
In #86 @michaelbarton created an example for json-schema. It allows us a way better handling of [validation errors](https://python-jsonschema.readthedocs.org/en/latest/errors/). The json-schema for our current input yaml would be:

```JSON
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "title": "Bioboxes short read assembler input file validator",
  "type": "object",
  "properties" : {

    "version" : {
      "type" :   "string",
      "pattern": "^0.9.\\d+$"  // Ensure the correct version is used via RE
    },

    "arguments" : {
      "type" : "array",
      "minItems" : 1,  # Limit number of items in input arguments
      "maxItems" : 2,
      "items" : [
        { "$ref": "#/definitions/fastq"    },  # Define definitions separately and
        { "$ref": "#/definitions/fragment" }   # then reference them
      ]
    }

  },
  "required": ["version", "arguments"],  # Required fields
  "additionalProperties" : false,        # No other fields
  "definitions": {

    "fastq" : {
      "type"                 : "object",
      "additionalProperties" : false,
      "required"             : ["fastq"],      # Ensure this a single key dictionary
      "properties" : {
        "fastq" : {
          "$ref": "#/definitions/values"}}},   # Use the values definition

    "fragment" : {
      "type"                 : "object",
      "additionalProperties" : false,
      "properties" : {
        "fragment_size" : {
          "$ref": "#/definitions/values"}}}, # Reuse the values definition second time

    "values" : {
      "type"        : "array",
      "uniqueItems" : true,
      "minItems"    : 1,
      "items"       : {
        "type"                 : "object",
        "additionalProperties" : false,
        "required"             : ["id", "value"],
        "properties"           : {
          "id"    : {},
          "type"  : {},
          "value" : {}
        }
      }
    }
  }
}
```

The usage of apropriate error messages is important for our [short read assembler validator](https://github.com/bioboxes/validator-short-read-assember/blob/feature/version/0.9.0/mount/features/assembler.feature). When the ray and soap examples pass the tests and the validator has enough reasonable tests, this issue can be closed. 
This is still maybe not the final schema library (see #86)

**Comment by pbelmann** [comment]
Examples should include [input validator](https://github.com/bioboxes/input-validator)

**Comment by pbelmann** [comment]
the correct yaml schema definition is now:
It is important to include the "oneOf" operator

```YAML
---
$schema: "http://json-schema.org/draft-04/schema#"
title: "Bioboxes short read assembler input file validator"
type: "object"
properties: 
    version: 
      type: "string"
      pattern: "^0.9.\\d+$"
    arguments: 
      type: "array"
      minItems: 1
      maxItems: 2
      items: 
        oneOf: 
          - 
            $ref: "#/definitions/fastq"
          - 
            $ref: "#/definitions/fragment"
required: 
    - "version"
    - "arguments"
additionalProperties: false
definitions: 
    fastq: 
      type: "object"
      additionalProperties: false
      required: 
        - "fastq"
      properties: 
        fastq: 
          $ref: "#/definitions/values"
    fragment: 
      type: "object"
      additionalProperties: false
      properties: 
        fragment_size: 
          $ref: "#/definitions/values"
    values: 
      type: "array"
      uniqueItems: true
      minItems: 1
      items: 
        type: "object"
        additionalProperties: false
        required: 
          - "id"
          - "value"
        properties: 
          id: {}
          type: {}
          value: {}
```

**Comment by avilella** [comment]
Thanks @pbelmann . Could we have a example schema and input files included
in the validator repo?

It would turn the validator repo into an initial complete example :-)

**Comment by pbelmann** [comment]
Do you mean the input-validator?
I think @michaelbarton wanted to create a generic validator that is independent of the biobox it uses. 
Because Binning/Profiling/Short-Read-Processor will have another schema definition. 
However I agree we should definitely include the schema in the rfc repo so the schema is for a biobox developer available. I think that is the subject of issue #103.
By the way I included the input validator in ray and soap.

**Comment by pbelmann** [comment]
Since we merged our new specification (0.9.0) for the assemblers and the examples (velvet, ray, soap) are implementing it I will close this issue.
